src/contracts/pods/EigenPod.sol

```diff
@@ -320,34 +299,127 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
+
+        require(validatorInfo.status == VALIDATOR_STATUS.INACTIVE,
+            "EigenPod.verifyCorrectWithdrawalCredentials: Validator must be inactive to prove withdrawal credentials");
+
+        require(validatorFields[BeaconChainProofs.VALIDATOR_WITHDRAWAL_CREDENTIALS_INDEX] == bytes32(_podWithdrawalCredentials()),
+            "EigenPod.verifyCorrectWithdrawalCredentials: Proof is not for this EigenPod");
+        /**
+        * Deserialize the balance field from the Validator struct.  Note that this is the "effective" balance of the validator
+        * rather than the current balance.  Effective balance is generated via a hystersis function such that an effective 
+        * balance, always a multiple of 1 ETH, will only lower to the next multiple of 1 ETH if the current balance is less
+        * than 0.25 ETH below their current effective balance.  For example, if the effective balance is 31ETH, it only falls to 
+        * 30ETH when the true balance falls below 30.75ETH.  Thus in the worst case, the effective balance is overestimating the
+        * actual validator balance by 0.25 ETH.  In EigenLayer, we calculate our own "effective reskated balance" which is a further pessimistic
+        * view of the validator's effective balance.
+        */
+        uint64 validatorEffectiveBalanceGwei = Endian.fromLittleEndianUint64(validatorFields[BeaconChainProofs.VALIDATOR_BALANCE_INDEX]);
+
+        // make sure the balance is greater than the amount restaked per validator
+        require(validatorEffectiveBalanceGwei >= REQUIRED_BALANCE_GWEI,
+            "EigenPod.verifyCorrectWithdrawalCredentials: ETH validator's balance must be greater than or equal to the restaked balance per validator");
+
+        // verify ETH validator proof
+        bytes32 latestBlockHeaderRoot = eigenPodManager.getBeaconChainStateRoot(oracleTimestamp);
+
+        // verify that the provided state root is verified against the oracle-provided latest block header
+       BeaconChainProofs.verifyStateRootAgainstLatestBlockHeaderRoot(proofs.beaconStateRoot, latestBlockHeaderRoot, proofs.latestBlockHeaderProof);
+
+
+        BeaconChainProofs.verifyValidatorFields(
+            validatorIndex,
+            proofs.beaconStateRoot,
+            proofs.validatorFieldsProof,
+            validatorFields
+        );
+
+        // set the status to active
+       validatorInfo.status = VALIDATOR_STATUS.ACTIVE;
+
+        emit ValidatorRestaked(validatorIndex);
+
+        //record validator's new restaked balance
+        validatorInfo.restakedBalanceGwei = _calculateRestakedBalanceGwei(validatorEffectiveBalanceGwei);
+
+        //record validatorInfo update in storage
+        _validatorPubkeyHashToInfo[validatorPubkeyHash] = validatorInfo;
+
+        // virtually deposit REQUIRED_BALANCE_WEI for new ETH validator
+        eigenPodManager.restakeBeaconChainETH(podOwner, validatorInfo.restakedBalanceGwei * GWEI_TO_WEI);
```

Comment:
we should be batching this update if many validators are proven

Reply:
oh whoops fixed.